src/engagement_db_to_analysis/engagement_db_to_analysis.py

```diff
@@ -72,66 +106,69 @@ def generate_analysis_files(user, google_cloud_credentials_file_path, pipeline_c
         engagement_db_datasets.extend(config.engagement_db_datasets)
     messages_map = get_messages_in_datasets(engagement_db, engagement_db_datasets, cache, dry_run)
 
-    messages_traced_data = _convert_messages_to_traced_data(user, messages_map)
+    channel_to_messages = _group_messages_by_channel_operator(messages_map, pipeline_config.analysis.channel_operators)
+    for channel, messages in channel_to_messages.items():
 
-    messages_traced_data = filter_messages(user, messages_traced_data, pipeline_config)
+        messages_traced_data = _convert_messages_to_traced_data(user, messages)
 
-    impute_codes_by_message(
-        user, messages_traced_data, analysis_dataset_configurations,
-        pipeline_config.analysis.ws_correct_dataset_code_scheme
-    )
+        messages_traced_data = filter_messages(user, messages_traced_data, pipeline_config)
 
-    messages_by_column = convert_to_messages_column_format(user, messages_traced_data, pipeline_config.analysis)
-    participants_by_column = convert_to_participants_column_format(user, messages_traced_data, pipeline_config.analysis)
+        impute_codes_by_message(
```

Comment:
Does this give the same results as main? If you impute codes after filtering by channel, then you might miss some stop codes.

Reply:
sure, I've resorted to imputing codes before filtering by channel.